### PR TITLE
Adapt ASN1_item_sign_ctx() for use with provided keypairs

### PR DESCRIPTION
**Pull request by levitte**
The mechanism to do this is to ask the signature operation for the DER
encoded AlgorithmIdentifier that corresponds to the combination of
signature algorithm and digest algorithm.

The provided DSA implementation is adapted to suit.

### CODE REVIEW

**Comment by levitte** [comment]
In the EXPERIMENTAL (#10797) test failure that triggered this is `test_req`

File: providers/implementations/signature/dsa.c [hunk 1]
```diff
@@ -116,24 +127,68 @@ static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
...
+     * We do so using EVP_MD_is_a(), and therefore need a name to NID
+     * map.
+     */
+    static const OSSL_ITEM name_to_nid[] = {
+        { NID_sha1,   OSSL_DIGEST_NAME_SHA1   },
+        { NID_sha224, OSSL_DIGEST_NAME_SHA224 },
+        { NID_sha256, OSSL_DIGEST_NAME_SHA256 },
+    };
+    size_t i;
+    int mdnid = -1;
```
Comment (slontis) [inline providers/implementations/signature/dsa.c]: Why use -1 instead of NID_undef

File: crypto/dsa/dsa_aid.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ENCODE_ALGORITHMIDENTIFIER_SHA2(sha256, 2);
+
+#define MD_CASE(name)                                                   \
+    case NID_##name:                                                    \
+        *len = sizeof(algorithmidentifier_##name##_der);                \
+        return algorithmidentifier_##name##_der
+
+const unsigned char *dsa_algorithmidentifier_encoding(int md_nid, size_t *len)
+{
+    switch (md_nid) {
```
Comment (slontis) [inline crypto/dsa/dsa_aid.c]: All SHA variants need to be in this table.. (including the truncated ones)
i.e...
Federal Information Processing Standard (FIPS) 186-4, Digital Signature Standard (DSS), specifies three NIST-approved digital signature algorithms: DSA, RSA, and ECDSA. All three are used to generate and verify digital signatures, in conjunction with an approved hash function specified in FIPS 180-4, Secure Hash Standard or FIPS 202, SHA-3 Standard: Permutation-Based Hash and Extendable-Output Functions.

**Comment by mattcaswell** [comment]
Travis failure appears relevant.

File: providers/implementations/signature/dsa.c (hunk 1 above)
Comment (levitte) [reply providers/implementations/signature/dsa.c]: Good question.

**Comment by levitte** [comment]
There will be corresponding work for ASN1_item_verify_ctx. That's a bit more involved because of horrid SM2 hackery in x_all.c. I'm pondering if I should make that a separate PR or include that in this PR. Thoughts?

File: crypto/dsa/dsa_aid.c (hunk 2 above)
Comment (levitte) [reply crypto/dsa/dsa_aid.c]: Hold on a minute...  This table here is about plain DSA.  Not RSA, not ECDSA.  Similar tables should and will be added in the RSA and ECDSA source (I do this in a sneaky way for RSA in #10557)

So for *this* table, the interesting OIDs are DSAWith{name}, {name}WithDSA, that sort of thing.  I haven't been able to find any that has SHA-3.  It's possible that I've starred myself blind.  Would you mind helping out?

File: providers/implementations/signature/dsa.c (hunk 1 above)
Comment (levitte) [reply providers/implementations/signature/dsa.c]: Fixed

**Comment by levitte** [comment]
> Travis failure appears relevant.

It should be fine now

File: crypto/dsa/dsa_aid.c (hunk 2 above)
Comment (slontis) [reply crypto/dsa/dsa_aid.c]: I cant find the OID's for the truncated ones (so they might not exist), but here are the others..
If we are supporting SHA3 it should be supported also for signing.
https://csrc.nist.gov/projects/computer-security-objects-register/algorithm-registration

Comment (levitte) [reply crypto/dsa/dsa_aid.c]: Thanks for the link.  Will add.

**Comment by slontis** [comment]
> It should be fine now

Rebuilt the failing travis tests (now that the error in master is fixed).
This should now only show real errors

File: crypto/dsa/dsa_aid.c (hunk 2 above)
Comment (levitte) [reply crypto/dsa/dsa_aid.c]: Fixed.

**Comment by slontis** [comment]
There is a memory leak..

File: include/openssl/core_names.h [hunk 3]
```diff
@@ -87,7 +87,10 @@ extern "C" {
 #define OSSL_DIGEST_PARAM_FLAGS      "flags"     /* ulong */
 
 /* Known DIGEST names (not a complete list) */
-#define OSSL_DIGEST_NAME_MD5 "MD5"
+#define OSSL_DIGEST_NAME_MD5            "MD5"
+#define OSSL_DIGEST_NAME_SHA1           "SHA1"
+#define OSSL_DIGEST_NAME_SHA224         "SHA2-224"
```
Comment (slontis) [inline include/openssl/core_names.h]: WHy are the other names missing?

File: providers/implementations/signature/dsa.c [hunk 4]
```diff
@@ -116,28 +127,84 @@ static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
...
+static int get_md_nid(const EVP_MD *md)
+{
+    /*
+     * Because the DSA library deals with NIDs, we need to translate.
+     * We do so using EVP_MD_is_a(), and therefore need a name to NID
+     * map.
+     */
+    static const OSSL_ITEM name_to_nid[] = {
+        { NID_sha1,   OSSL_DIGEST_NAME_SHA1   },
+        { NID_sha224, OSSL_DIGEST_NAME_SHA224 },
```
Comment (slontis) [inline providers/implementations/signature/dsa.c]: Also missing here?

File: providers/implementations/signature/dsa.c [hunk 5]
```diff
@@ -116,28 +127,84 @@ static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
...
+        goto error;
+
+    memcpy(pdsactx->aid, algorithmidentifier, algorithmidentifier_len);
+    pdsactx->aid_len = algorithmidentifier_len;
 
     if (!EVP_DigestInit_ex(pdsactx->mdctx, md, NULL))
-        return 0;
+        goto error;
 
     return 1;
```
Comment (slontis) [inline providers/implementations/signature/dsa.c]: I assume the memory leak is related to the return here..

**Comment by paulidale** [comment]
Is Travis relevant here?  (fuzz test failure).

**Comment by slontis** [comment]
> Is Travis relevant here? (fuzz test failure).

Its a memory leak picked up by asan

File: providers/implementations/signature/dsa.c (hunk 4 above)
Comment (levitte) [reply providers/implementations/signature/dsa.c]: Fixed

File: include/openssl/core_names.h (hunk 3 above)
Comment (levitte) [reply include/openssl/core_names.h]: Fixed

File: providers/implementations/signature/dsa.c (hunk 5 above)
Comment (levitte) [reply providers/implementations/signature/dsa.c]: Kinda sorta.  `md` could remain unfreed under certain circumstances.  Anyway, fixed, I hope.

**Comment by levitte** [comment]
Time for a re-review

File: providers/implementations/signature/dsa.c [hunk 6]
```diff
@@ -116,28 +127,88 @@ static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
...
-    pdsactx->md = md;
+    pdsactx->md = EVP_MD_fetch(pdsactx->libctx, mdname, props);
+    algorithmidentifier =
+        dsa_algorithmidentifier_encoding(get_md_nid(pdsactx->md),
+                                         &algorithmidentifier_len);
+
+    if (algorithmidentifier == NULL)
+        goto error;
+
     pdsactx->mdsize = EVP_MD_size(md);
```
Comment (slontis) [inline providers/implementations/signature/dsa.c]: pdsactx->md

File: crypto/dsa/dsa_aid.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#include "crypto/dsa.h"
+
+#define ASN1_SEQUENCE 0x30
+#define ASN1_OID 0x06
+
+/* dsaWithSHA OIDs are of the form: (1 3 14 3 2 |n|) */
+#define ENCODE_ALGORITHMIDENTIFIER_SHA(name, n)                         \
+    static const unsigned char algorithmidentifier_##name##_der[] = {   \
+        ASN1_SEQUENCE, 0x06,                                            \
+          ASN1_OID, 0x09, 1 * 40 + 3, 14, 3, n,                         \
```
Comment (slontis) [inline crypto/dsa/dsa_aid.c]: This oid looks like it has the wrong length and is missing a 2 (also remove the trailing ',')

File: providers/implementations/signature/dsa.c (hunk 6 above)
Comment (levitte) [reply providers/implementations/signature/dsa.c]: Fixed

File: crypto/dsa/dsa_aid.c (hunk 7 above)
Comment (levitte) [reply crypto/dsa/dsa_aid.c]: Fixed

File: crypto/dsa/dsa_aid.c [hunk 8]
```diff
@@ -18,15 +18,15 @@
 /* dsaWithSHA OIDs are of the form: (1 3 14 3 2 |n|) */
 #define ENCODE_ALGORITHMIDENTIFIER_SHA(name, n)                         \
     static const unsigned char algorithmidentifier_##name##_der[] = {   \
-        ASN1_SEQUENCE, 0x06,                                            \
-          ASN1_OID, 0x09, 1 * 40 + 3, 14, 3, n,                         \
+        ASN1_SEQUENCE, 0x07,                                            \
+          ASN1_OID, 0x09, 1 * 40 + 3, 14, 3, 2, n                       \
```
Comment (slontis) [inline crypto/dsa/dsa_aid.c]: Is the 0x09 supposed to be 0x05
```suggestion
          ASN1_OID, 0x09, 1 * 40 + 3, 14, 3, 2, n                       \
```

Comment (levitte) [reply crypto/dsa/dsa_aid.c]: Yes.  Fixed.  Thanks for being eagle-eyed :-)

File: crypto/asn1/a_sign.c [hunk 9]
```diff
@@ -156,11 +157,51 @@ int ASN1_item_sign_ctx(const ASN1_ITEM *it,
...
+            if (d2i_X509_ALGOR(&algor1, &pp, aid_len) == NULL) {
+                ASN1err(ASN1_F_ASN1_ITEM_SIGN_CTX, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+        }
+
+        if (algor2 != NULL) {
+            const unsigned char *pp = aid;
+
+            if (d2i_X509_ALGOR(&algor2, &pp, aid_len) == NULL) {
```
Comment (slontis) [inline crypto/asn1/a_sign.c]: Just checking, this code is sitting outside the provider?

**Review by slontis — APPROVED** [review]

File: crypto/asn1/a_sign.c (hunk 9 above)
Comment (levitte) [reply crypto/asn1/a_sign.c]: I believe we have informally abolished everything in crypto/asn1 from the fips module, so yes.

**Comment by levitte** [comment]
Merged

d5aef5946bd9b113623ad778114768585a1f7a02 Adapt ASN1_item_sign_ctx() for use with provided keypairs
505b41fc5a7a3cb255d2f62cf4902a1a5c1db2dd PROV: Adapt the DSA signature implementation to provide Algorithmidentifiers
0cb3f4f985aa52688ddc90075f480ba612d34fe3 test_evp_extra_test.c: don't rely on exact parameter position

File: crypto/dsa/dsa_aid.c [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha224, 1);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha256, 2);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha384, 3);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha512, 4);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha3_224, 5);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha3_256, 6);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha3_384, 7);
+ENCODE_ALGORITHMIDENTIFIER_SHAx(sha3_512, 8);
```
Comment (paulidale) [inline crypto/dsa/dsa_aid.c]: This is perfect for FIPS.  It isn't so good if dsa-with-XXX exists for other digests.  Probably more of an issue for ECDSA than here.

**Pull request closed without merging**